tree/ntuple/v7/src/RClusterPool.cxx

```diff
@@ -105,51 +104,50 @@ void ROOT::Experimental::Detail::RClusterPool::ExecUnzipClusters()
 void ROOT::Experimental::Detail::RClusterPool::ExecReadClusters()
 {
    while (true) {
-      std::vector<RReadItem> readItems;
-      std::vector<RCluster::RKey> clusterKeys;
-      std::int64_t bunchId = -1;
+      std::deque<RReadItem> readItems;
```

Comment:
Is that legal for non-`static` class members?  And if it is, doesn't that defeat the purpose of communication between threads?

Reply:
> Is that legal for non-static class members?

Class data members can not be thread local.

Here I am proposing a function thread local (where the function is member of class).  The thread local `readItems` would be seen only by the current thread (same as a stack object) but would be re-used the next time the function is re-entered (by the same thread).   We would (of course) have to remember to 'clear' it as the end of the function (so that as the next use it has some reserved memory but is 'empty').

Reply:
Ah, sorry, I misinterpreted the context of the suggestion :cold_sweat:.  However, making `readItems` thread-local does not address the issue in the PR description, i.e.
by the time the thread awakes after a `fCvHasReadWork.wait()`, `readItems` is a empty default-initialized container.  Therefore a `std::swap(readItems, fReadQueue)` causes `fReadQueue` to be an empty container with no allocated memory; further insertions from the "producer" thread need to allocate again.

Reply:
> readItems is a empty default-initialized container. 

Yes, the first time the thread is used.  Is `ROOT::Experimental::Detail::RClusterPool::ExecUnzipClusters()` always executed on a new thread? (humm it should not be the case when we use task/tbb).

Reply:
`ExecReadClusters()` (same for `ExecUnzipClusters()`) is the entry point for a raw `std::thread` that has the same lifetime as the `RClusterPool` itself, which in principle should also match the lifetime of a page source, e.g. the `RPageSourceFile` constructor initializes a new `RClusterPool` whose life will span until the destruction of the `RPageSourceFile` instance.  The read thread is expected to be blocked in I/O most of the time.

Reply:
Fair enough, so indeed the thread local does not help.  To understand whether we have in the PR is good enough, let me try to summarize the options:
(1) Use std::swap on collection.  Pro: reduce the time-length of the critical section.  Cons: increase other critical section by requirement reallocation of the deque.
(2) Use std::swap on each elements. Pro: avoid re-allocation while filling fReadItems. Cons: critical section is time-wise longer.
I.e. It is no clear to me which of (1) and (2) would be the more performant.  If there are similar, I would go for the simpler implementation (i.e. using std::swap on the collection).

If (2) is longer, we could try to find other ways to still avoid the reallocation while also using std::swap on the collection.  For example, there could be a collection of 'spare' `std::deque` that a thread would check out from to swap with fReadItem, use its content and put it back (hence 'saving' the allocation) on the set of `spare`.

Reply:
Updates have been pushed; please, take a look if you have some spare time.  Thanks!